lib/Compilation.js

```diff
@@ -644,7 +901,10 @@ class Compilation extends Tapable {
 		) {
 			throw new Error("Parameter 'dependency' must be a Dependency");
 		}
-
+		// @ts-ignore
+		// TODO: Not sure how to handle the
+		// dependency.constructor error,
+		// maybe I typed it wrong
```

Comment:
Turns out lib.d.ts defines `Object.prototype.constructor: Function`. In order for there not to be a type error, it would have to be `new (...args: any[]) => this`. So it's not your fault, it's ours.

If you add a cast, you can work around this:

`/** @type {new (...args: any[]) => Dependency} */(dependency.constructor)`